[Draft] New addon `find-references` (find reference list window)

OriLiMu opened the pull request:
<!-- Which issue(s) does this pull request fix or resolve? If there aren't any, please submit one first unless this is a hotfix or minor string update. -->

Resolves #7594 

### Changes
Base on the addon **find-bar**
1. add a float window at right of the editor
2. you can search for the variable name or event block or self defined function 
3. result list shown as images. with start block and directly call block. for example if there are two blocks stacked(lets say block A and block B). And the second block B call the  var **X**. Then you search for the reference for var **X**. A would be the  start block. B is the directly call block. 


### Reason for changes

for #7594

### Tests

Just tested on chrome for now

Commit:
first commit

Commit:
update to ref

Commit:
Revert "update to ref"

This reverts commit db2e0b9685f4c4f037fe5854dacbd0e21883a679.

Commit:
add jsdoc

Commit:
Update userscript.js

Commit:
add jsdoc

Commit:
Update userscript.js

Commit:
input-bar width changed

Commit:
format mainly

Commit:
Update userscript.js

Commit:
drag window done

Commit:
set back the  ujs

Commit:
item test done

Commit:
hover done

Commit:
update

Commit:
revert find bar

Commit:
update

Commit:
update

Commit:
right near done

Commit:
round corner done

Commit:
add find bar done

Commit:
find -bar fix

Commit:
update

Commit:
popout done

Commit:
bar shown in float window

Commit:
basic color and position fixed

Commit:
basic adjust done

Commit:
msg done

Commit:
ul set done

Commit:
svg insert done

Commit:
function amend done

Commit:
insert svg done but size wrong

Commit:
remove default css

Commit:
add top svg done

Commit:
row set done

Commit:
partly done get top block

Commit:
align block

Commit:
some kind make it

Commit:
update

Commit:
add the button

Commit:
eye effect done

Commit:
Update userstyle.css

Commit:
update

Commit:
update

Commit:
css on label

Commit:
dropdown show

Commit:
the sa-fr-input height is changed

Commit:
result size adjust done

Commit:
Update userstyle.css

OriLiMu committed:
adjust size

OriLiMu committed:
add ref_list li padding

Commit:
Merge branch 'master' of github.com:woshichuanqilz/ScratchAddons

Samq64 commented on .gitignore:
```diff
@@ -34,4 +34,6 @@ ehthumbs_vista.db
-node_modules/
\ No newline at end of file
+node_modules/
+addons/find-bar/userscript.js
+addons/find-references/userstyle_1.css
```
```suggestion
```
Please don't modify `.gitignore`, the old files are still in git's history. If you want to keep them around locally just don't commit those files or move them to the desktop.

Samq64 commented on addons/addons.json:
```diff
@@ -182,5 +181,6 @@
   "editor-compact",
-  "editor-theme3"
+  "editor-theme3",
+  "find-references"
```
New addons usually go above themes. Have you checked this order is necessary?

woshichuanqilz replied:
I don't know that this order matters. I was wondering why the colorscheme changed not work on my addon. 
Seems this is the reason.

Samq64 commented on addons/find-references/blockly/Utils.js:
```diff
@@ -0,0 +1,357 @@
+
+    workspace.scrollbar.set(sx, sy);
+
+    /*
+              let blocklySvg = document.getElementsByClassName('blocklySvg')[0];
+              let blocklyBlockCanvas = blocklySvg.getElementsByClassName('blocklyBlockCanvas')[0];
+              let transform = blocklyBlockCanvas.transform.baseVal.getItem(0);
+              let scale = blocklyBlockCanvas.transform.baseVal.getItem(1);
+
+              let transformMatrix = transform.matrix;
+              let scaleMatrix = scale.matrix;
+
+              console.log('Transform - getMetrics', s);
+              console.log('sx, sy: ', sx, sy);
+              console.log('left, top: ', view.left, view.top);
+              console.log('contentLeft, right:', s.contentLeft, s.contentTop);
+              console.log('transform, scale matrix: ', transformMatrix, scaleMatrix);
+  */
```
```suggestion
```

Samq64 reviewed (commented):
Could the changes in unrelated files be removed?

Samq64 commented:
Also if this is intended to replace `find-bar` either the addon ID should stay the same or all of `find-bar`'s code should be deleted.

OriLiMu committed:
gitignore cant be changed

Co-authored-by: Samq64 <81489795+Samq64@users.noreply.github.com>

woshichuanqilz commented:
> Also if this is intended to replace `find-bar` could you keep the addon ID the same or remove all of `find-bar`'s code?

I use different addon id for now. `find-bar` is simple to use and with neat lookinig. I think it cant be replaced now .

Samq64 commented:
> I see there is a conflict with find bar in my branch. My modification on find-bar can be totally ignored.
>
> Need I to delete the unrelated file and make a pr again? Not quite familiar with the work flow

You can just commit and push more changes. We can also help resolve those issues for you.

Samq64 committed:
Undo unnecessary diffs

Commit:
fix base on pr

1. remove blockly folder directly use find-bar's
2. create svg_utils.js related to svg function
3. revert the change in .gitignore

Commit:
change addon order in addons.js

woshichuanqilz commented:
> > I see there is a conflict with find bar in my branch. My modification on find-bar can be totally ignored.
> > Need I to delete the unrelated file and make a pr again? Not quite familiar with the work flow
> 
> You can just commit and push more changes. We can also help resolve those issues for you.

A little confused about the git command right now. 

All my changes is committed. 

When I run  ` git push origin fr-branch` I got this error.  Afraid of make mistake again. What command should I use now ? 
```
To github.com:woshichuanqilz/ScratchAddons.git
 ! [rejected]          fr-branch -> fr-branch (non-fast-forward)
error: failed to push some refs to 'github.com:woshichuanqilz/ScratchAddons.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. If you want to integrate the remote changes,
hint: use 'git pull' before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

```

Samq64 commented:
You can use [GitHub desktop](https://github.com/apps/desktop) do avoid dealing with the command line stuff. If there are conflicts you can try resolving them in VSCode if you have that installed.

Commit:
Merge branch 'master' of github.com:woshichuanqilz/ScratchAddons into fr-branch

Commit:
Revert "Merge branch 'master' of github.com:woshichuanqilz/ScratchAddons into fr-branch"

This reverts commit 113eea7e7966f6ca475bbd018643083bfe162d52, reversing
changes made to b4ee68a64f0cbcadf4b441a20c1a052ca471f13e.

OriLiMu committed:
Merge branch 'fr-branch' of github.com:woshichuanqilz/ScratchAddons into fr-branch

Samq64 commented:
Maybe it won't work any more but try merging upstream/master into your branch.

woshichuanqilz commented:
already merged it I think. can you see that?

Samq64 ***@***.***> 于 2024年7月20日周六 04:34写道：

> Maybe it won't work any more but try merging upstream/master into your
> branch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ScratchAddons/ScratchAddons/pull/7664#issuecomment-2240066620>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWFU65YEV23YR3743M5LKDZNFZ63AVCNFSM6AAAAABLDYLWOKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENBQGA3DMNRSGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Samq64 committed:
Merge branch 'master' into pr/7664

Samq64 committed:
Revert "Merge branch 'master' into pr/7664"

This reverts commit 05819cf04eca6df57edfa680ae9b0058b08967cc, reversing
changes made to b686adee7e2da356cb6a5a2578405a56b6eec8d4.

Samq64 committed:
Reapply "Merge branch 'master' of github.com:woshichuanqilz/ScratchAddons into fr-branch"

This reverts commit baf2fbb7560f15a39ef9f9da1dc352a02ffd5595.

Samq64 committed:
Attempt to re-merge from upstream/master

This reverts commit b4ee68a64f0cbcadf4b441a20c1a052ca471f13e.

woshichuanqilz commented on addons/find-bar/userscript.js:
I change the file by accident, I don't know how to ignore this change

OriLiMu reviewed (commented):
I change the file by accident, I don't know how to ignore this change
Need I to commit again?

OriLiMu committed:
revert find-bar userscript which is modified by accidently

OriLiMu committed:
Merge branch 'fr-branch' of github.com:woshichuanqilz/ScratchAddons into fr-branch

woshichuanqilz commented:
what is the meaning of this :
![image](https://github.com/user-attachments/assets/6372de15-0875-4167-b70b-02bb4ce9a5f7)

I submit the change to my repo works fine. 
Need I to run the command `git push origin HEAD:<name-of-remote-branch> from my side ?

Joeclinton1 commented:
> what is the meaning of this : ![image](https://private-user-images.githubusercontent.com/11295355/350700565-6372de15-0875-4167-b70b-02bb4ce9a5f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE1Nzc4MjEsIm5iZiI6MTcyMTU3NzUyMSwicGF0aCI6Ii8xMTI5NTM1NS8zNTA3MDA1NjUtNjM3MmRlMTUtMDg3NS00MTY3LWI3MGItMDJiYjRjZTlhNWY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIxVDE1NTg0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyOGM3MTU2ODJjYmIwZGQ0MjY0MDQ3ZDk4YjUxYWZmMDQyNDMyNzhjYzZkMDhhNDMwMTFhOGJlNjgyYjFmNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZQpIVGxWbyAsdWUP-eY4MvEbAzDMR-tpCIpTVletDas)
> 
> I submit the change to my repo works fine. Need I to run the command `git push origin HEAD: from my side ?

Get a GUI with a branch visualizer to make this easier on your self. Also before pushing Github PR's you really should familiarize yourself with Git.

See https://graphite.dev/guides/you-are-not-currently-on-a-branch for resolving the error. This link also includes a best practices section for avoiding the error in future which might be useful.

DNin01 commented:
> See https://graphite.dev/guides/you-are-not-currently-on-a-branch for resolving the error. This link also includes a best practices section for avoiding the error in future which might be useful.

I think that was a GitHub Actions error. It just means some checks failed.

The error occurs when our workflow for code formatting tries to make changes to another user's repository.
&nbsp;

To have your code automatically formatted, you can [enable Actions](https://github.com/woshichuanqilz/ScratchAddons/actions) on _your_ fork of the repository.

OriLiMu committed:
remove unrelated code

Commit:
Format code

OriLiMu committed:
remove un need comment

OriLiMu committed:
Merge branch 'fr-branch' of github.com:woshichuanqilz/ScratchAddons into fr-branch

OriLiMu committed:
don't click small pane

Commit:
Format code

OriLiMu committed:
adjust position of svg

OriLiMu committed:
remove comments

OriLiMu committed:
Merge branch 'fr-branch' of github.com:woshichuanqilz/ScratchAddons into fr-branch

woshichuanqilz commented:
I don't know what to do on this. Except this #7659. 
May I click the ready for review button ?

OriLiMu committed:
Merge branch 'master' into fr-branch
